### ISSUE TITLE
Theoretically exact sine/cosine method

### ISSUE THREAD

**Issue by nslay** [Morwenn/static_math#19]
Very cool library! I found it mentioned in a constexpr stackexchange question. I like to dabble with low level math like this and I've derived a theoretically exact way to calculate sine/cosine by playing games with the half-angle formulae and bits. I've provided the derivation and a (probably poorly written) C++14 constexpr version (requires constexpr abs, sqrt, sgn, and modf). You might already know this trickery. I can't promise the C++14 version is correct (Clang 6 is happy with it) or if MSVC will tolerate it.

I'm not suggesting you use this. Your implementation using Taylor series (if I'm not mistaken) is probably a lot faster since this formulation needs to examine at most the number of mantissa bits (e.g. 52 for double precision)! 

I think there is a name for this method. I don't remember off the top of my head. But good fun! 

Thanks for the library!

Here's the demo code... I can just use your static math library from now on!
[demo.txt](https://github.com/Morwenn/static_math/files/2164422/demo.txt)

And below are notes I wrote myself if I ever look at this code again in the distant future.

NOTE: I didn't intend for the LaTeX pictures to automatically display.

**Explanation**
![explanation](https://user-images.githubusercontent.com/6295986/42294777-bcd61f46-7fb1-11e8-9025-d207e3fcf387.png)

**Pseudo Code**
![pseudocode](https://user-images.githubusercontent.com/6295986/42294779-be956058-7fb1-11e8-8e5f-d89243d8f6b4.png)

**Comment by ratchetfreak** [comment]
it's not going to be 100% exact though as rounding errors will accumulate and lose you up to log2(steps) = ceil(log2(54)) = 7 bits.

**Comment by nslay** [comment]
Indeed, it's only exact on paper. But it's exact on paper for finitely represented binary numbers using a finite number of operations. Contrast this with infinite series that need infinite terms to be exact on paper.

This kind of algorithm is seemingly related to CORDIC algorithms (that's the name I couldn't remember), although my derivation is different than CORDIC described on Wikipedia. That one appears to produce 1 bit of the output per iteration which I think is far more interesting than what I've derived.

https://en.wikipedia.org/wiki/CORDIC

Looks like CORDIC algorithms exist for most mathematical functions. Such algorithms might be good candidates for constexpr math.

**Comment by Morwenn** [comment]
Monthly reminder that this library was a toy project and that I know next to nothing to floating point math internals and non-trivial math in general. The exponential, logarithmic, trigonometric and hyperbolic functions have been contributed by other people who are more knowledgeable in those areas than I am :p

Algorithms with fewer rounding errors do sound interesting but I am hardly in a position to judge what is actually better.
